Upgrade DataStax cassandra driver to 4.9.0 [full build]

porunov opened the pull request:
Mostly based on #1733 .

This PR adds 2 DataStax drivers into `janusgraph-cql`:
- DataStax 4.9.0 is the main driver which is now used.
- DataStax 3.10.2 is the driver which is now used for Hadoop. `CqlBinaryRecordReader` which is used for Hadoop jobs is still using DataStax Cassandra Driver 3.x.y.

Notice, even so this PR adds 2 different versions of DataStax Cassandra Driver in the same module, those drivers are binary incompatible and share different classpath, thus they don't conflict with each other and can be used simultaneously. 
Later we can move CQL Hadoop formats from `janusgraph-cql` to a new module like `janusgraph-hadoop-cql` or something like that. It will allow to include only 4.x version of the driver into classpath for users who use `janusgraph-cql` but don't use `janusgraph-hadoop-cql`.

Tests passed:
- Custom integration tests. Storage backend: ScyllaDB 4.2.0 (cql driver). Index backend: ElasticSearch 7.6.0 (es driver).
- Distribution build. `janusgraph.sh` starts Cassandra and ElasticSearch. `janusgraph-server` successfully connects to storage backends via cql driver.
- janusgraph-full distribution successfully logs different metrics via JMX (https://github.com/JanusGraph/janusgraph/issues/1087#issuecomment-534513530), console, slf4j.
- Embedded janusgraph successfully logs different metrics via JMX, console, slf4j. Also, tried to log new CQL Session Level and Node Level metrics via [JMX exporter](https://github.com/prometheus/jmx_exporter). New CQL Session Level and Node Level metrics to enable are provided via `storage.cql.metrics.session-enabled` and `storage.cql.metrics.node-enabled` respectively. 
- All TinkerPop tests passed. Executed with `mvn clean install -Dtest.skip.tp=false -DskipTests=true`.
- TravisCI `[full build]` passed.

Compared performance of `master` branch (f5001e6210baa15bc136f3224ba1e83521782cba) and this PR (65cad01828f1a455cc51435f8efaf309111c0768). The performance is fairly the same. I tried several times to execute custom integration tests which showed that the PR branch in most cases runs about 1%-1.5% faster than `master` branch. I tried to execute CQL TP tests 2 times for each branch which showed again that this PR runs 1%-1.5% faster (in average 4:20 h for this PR and 4:25 h for master branch). The performance improvement most likely achieved because of a slightly more optimal `CQLPagingIterator` implementation and most likely not directly related to driver upgrade. Overall I would say that the performance is the same but I didn't get enough statistics for real comparison. The custom integration tests were executed on  ScyllaDB 4.2.0 on the same machine with JanusGraph in embedded mode with no replication.   

Changelog: https://docs.datastax.com/en/developer/java-driver/4.9/changelog/

Fixes #1510
Potentially solves #2254 for `master` branch (`0.6.0` release)

Signed-off-by: Oleksandr Porunov <alexandr.porunov@gmail.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[doc only]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

porunov commented:
There are many dependency conflicts. Not sure I will be able to resolve them. Maybe will try it later. If someone wants to get this PR and work on it feather, feel free to do it.

porunov commented:
Instead of excluding dropwizard, I've excluded dropwizard 4.0.5 from datastax driver. Looks like it doesn't complain about dropwizard anymore.
Tests are still failing. Will need to research why it happens.

porunov commented:
@marco-scoppetta if you could provide any guidance or review this PR, it would be very helpful. It based purely on your PR but contains some fixes. I guess I am missing something still.

farodin91 commented:
@porunov I did work on bump of Solr as best as i can. I was more like try and error upgrade. 
I hope we are able to add support for Testcontainers as soon as possible (tried it multiple times).

farodin91 commented:
@porunov you can now fix the issue with dropwizard.

smcquillan commented:
@porunov I made some progress while looking at the following exception from the travis build:
```
com.datastax.oss.driver.api.core.AllNodesFailedException: Could not reach any contact point, make sure you've provided valid addresses (showing first 1 nodes, use getAllErrors() for more): Node(endPoint=localhost/127.0.0.1:32770, hostId=null, hashCode=ef6be2): [com.datastax.oss.driver.api.core.connection.ConnectionInitException: [JanusGraph Session|control|id: 0xead65598, L:/127.0.0.1:36916 - R:localhost/127.0.0.1:32770] Protocol initialization request, step 1 (OPTIONS): failed to send request (javax.net.ssl.SSLException: SSLEngine closed already)]
```
I was able to reproduce this locally using:
```
 mvn verify --projects janusgraph-cql -Pcassandra3-murmur-ssl -Dcassandra.docker.version=3.11.4 -Dtest=**/diskstorage/cql/CQLStoreTest.java 
```
And found that adding the following line to `CQLStoreManager::initializeSession()`
```java
            configLoaderBuilder.withString(DefaultDriverOption.SSL_HOSTNAME_VALIDATION, "false");
```
Resolved the problem. Though I doubt we can just hard code `SSL_HOSTNAME_VALIDATION` to false as it'll need to be a configuration option. 

There's still another issue though, because I reproduced using just `-Pcassandra3-murmur-ssl` but the travis tests use `-Pcassandra3-murmur-client-auth` and when running that locally I'm now seeing a new error about bad certificates. I'll try and investigate that one further later.

smcquillan commented:
@porunov I'm now able to run those tests locally using the client authentication with:
```
 mvn verify --projects janusgraph-cql -Pcassandra3-murmur-client-auth -Dcassandra.docker.version=3.11.4 -Dtest=**/diskstorage/cql/CQLStoreTest.java
```
The change needed to stop the bad certificate error was to add in the keystore password and location. In `CQLStoreManager` where your PR reads:
```java
        if (configuration.get(SSL_ENABLED)) {
            configLoaderBuilder
                .withClass(DefaultDriverOption.SSL_ENGINE_FACTORY_CLASS, DefaultSslEngineFactory.class)
                .withString(DefaultDriverOption.SSL_TRUSTSTORE_PATH, configuration.get(SSL_TRUSTSTORE_LOCATION))
                .withString(DefaultDriverOption.SSL_TRUSTSTORE_PASSWORD, configuration.get(SSL_TRUSTSTORE_PASSWORD));
        }
```
To get the tests running locally I changed this to
```
        if (configuration.get(SSL_ENABLED)) {
            configLoaderBuilder
                .withClass(DefaultDriverOption.SSL_ENGINE_FACTORY_CLASS, DefaultSslEngineFactory.class)
                .withString(DefaultDriverOption.SSL_KEYSTORE_PATH, configuration.get(SSL_KEYSTORE_LOCATION))
                .withString(DefaultDriverOption.SSL_KEYSTORE_PASSWORD, configuration.get(SSL_KEYSTORE_KEY_PASSWORD))
                .withString(DefaultDriverOption.SSL_TRUSTSTORE_PATH, configuration.get(SSL_TRUSTSTORE_LOCATION))
                .withString(DefaultDriverOption.SSL_TRUSTSTORE_PASSWORD, configuration.get(SSL_TRUSTSTORE_PASSWORD));
        
            configLoaderBuilder.withString(DefaultDriverOption.SSL_HOSTNAME_VALIDATION, "false");
        }
```
Hopefully that'll be enough to get those tests running the the travis build.

smcquillan commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:
```diff
@@ -450,7 +473,7 @@ public CQLPagingIterator(final int pageSize, Supplier<Statement> statementSuppli
 
         @Override
         public boolean hasNext() {
-            return !currentResultSet.isExhausted();
+            return !currentResultSet.isFullyFetched();
```
@porunov Looks like `isExhausted()` was removed in driver 4.x but it wasn't replaced by `isFullyFetched()` as that also existed in version 3.x. From looking at the implementation of `currentResultSet.one()` I assume that here we should be writing:
```java
            return currentResultSet.iterator().hasNext();
```
That seems to get a few more tests passing.

smcquillan commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLStoreManager.java:
```diff
@@ -148,9 +82,7 @@
-    private Cluster cluster;
-    @Resource
-    private Session session;
+    private final CqlSession session;
```
@porunov Think I finally now understand the reason why some of the tests are failing with the Mockito exception
```
org.mockito.exceptions.misusing.UnnecessaryStubbingException
```
And it's because the `session` member variable was made `final` and this prevents the `@InjectMocks` annotation in `CQLStoreTest` from injecting the mock session instance which the tests are expecting. When I remove the `final` qualifier from the session the tests pass for me.

marco-scoppetta commented:
Hey @porunov see this comment: https://github.com/JanusGraph/janusgraph/pull/1733#issuecomment-632832050

As I have mentioned there, I think the interesting part of CQLStoreManager which is worth knowing is what I had implemented [here](https://github.com/graknlabs/grakn/blob/master/graph/diskstorage/cql/CQLStoreManager.java#L230) (line 230 to 243) (also links to CQL discussion group). Setting those timeouts to 0 or low values o make them configurable will yield immediate gains.

Specially for your CQL tests you have a custom `CachingCQLStoreManager` which only exists because tests were timing out, and that's because the `close()` method of new CQL driver by default waits 2 seconds before shutting down, if you were to set the timeouts to 0 you wouldnt need the custom class.

Everything else looks good I think

smcquillan commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLStoreManager.java:
```diff
@@ -227,114 +159,71 @@ public CQLStoreManager(final Configuration configuration) throws BackendExceptio
-            }
+            configLoaderBuilder
+                .withClass(DefaultDriverOption.SSL_ENGINE_FACTORY_CLASS, DefaultSslEngineFactory.class)
+                .withString(DefaultDriverOption.SSL_KEYSTORE_PATH, configuration.get(SSL_KEYSTORE_LOCATION))
+                .withString(DefaultDriverOption.SSL_KEYSTORE_PASSWORD, configuration.get(SSL_KEYSTORE_KEY_PASSWORD))
```
@porunov Build is  failing on `cassandra3-murmur-ssl` with:
```
java.lang.IllegalArgumentException: Invalid configuration value for [root.storage.cql.ssl.keystore.location]:
```
This is because we're setting the key store options when client auth has not been enabled. So these options need moved out to something like:
```java
            if(configuration.get(SSL_CLIENT_AUTHENTICATION_ENABLED)) {
            	configLoaderBuilder.withString(DefaultDriverOption.SSL_KEYSTORE_PATH, configuration.get(SSL_KEYSTORE_LOCATION))
                                   .withString(DefaultDriverOption.SSL_KEYSTORE_PASSWORD, configuration.get(SSL_KEYSTORE_KEY_PASSWORD));
            }
```

smcquillan commented:
@porunov Two of the remaining failing tests are `testClearStorage` and `testGlobalGraphConfig` which often fail with timeouts, though I have occasionally seen them pass when running locally.

I think there might be a connection with the suggestion from @marco-scoppetta as when I set those timeouts to zero the tests pass pretty much every time.

porunov commented:
Thank you @marco-scoppetta and @smcquillan ! I will take a look at it soon

porunov commented:
The tests finally passed. That's a good sign. 
Early next week I plan to:
- Review the changes in this PR (in particularly configuration options)
- Test the PR internally
- Run TinkerPop tests

As always any reviews / comments / tests are very welcome!

smcquillan commented:
Good to hear we've got the tests over the line :)

I'm off on vacation from tomorrow so won't be online for a couple of weeks, but from what I've seen of the changes they look good to me.

porunov commented:
TinkerPop tests fail for this PR with:
```
[ERROR]   StarGraphTest.shouldValidateThatOriginalAndStarVerticesHaveTheSameTopology Multiple Failures (2 failures)
	org.testcontainers.containers.ContainerLaunchException: Container startup failed
	java.lang.NullPointerException: <no message>
```

I will take a look at it later

smcquillan commented:
@porunov When running the tests locally I found the also failed with the same error message. However, I then realised that it was because I'd not started docker. With Docker running  and using ` mvn clean install   --projects janusgraph-cql   -Dtest.skip.tp=false -DskipTests=true` most of the tests passed:
```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/Users/smcquill/.m2/repository/org/slf4j/slf4j-log4j12/1.7.30/slf4j-log4j12-1.7.30.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/Users/smcquill/.m2/repository/ch/qos/logback/logback-classic/1.1.3/logback-classic-1.1.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/Users/smcquill/.m2/repository/org/slf4j/slf4j-log4j12/1.7.30/slf4j-log4j12-1.7.30.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/Users/smcquill/.m2/repository/ch/qos/logback/logback-classic/1.1.3/logback-classic-1.1.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
[INFO] Running org.janusgraph.blueprints.cql.process.CQLComputerTest
[ERROR] Tests run: 773, Failures: 4, Errors: 0, Skipped: 142, Time elapsed: 12,503.042 s <<< FAILURE! - in org.janusgraph.blueprints.cql.process.CQLComputerTest
[ERROR] shouldRespectThreadInterruptionInVertexStep[expectInterruption{g_V_out}]  Time elapsed: 18.456 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: is <true>
     but: was <false>
	at org.apache.tinkerpop.gremlin.process.traversal.TraversalInterruptionComputerTest.shouldRespectThreadInterruptionInVertexStep(TraversalInterruptionComputerTest.java:100)

[ERROR] shouldRespectThreadInterruptionInVertexStep[expectInterruption{g_V_in}]  Time elapsed: 18.529 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: is <true>
     but: was <false>
	at org.apache.tinkerpop.gremlin.process.traversal.TraversalInterruptionComputerTest.shouldRespectThreadInterruptionInVertexStep(TraversalInterruptionComputerTest.java:100)

[ERROR] shouldRespectThreadInterruptionInVertexStep[expectInterruption{g_V_inE}]  Time elapsed: 18.159 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: is <true>
     but: was <false>
	at org.apache.tinkerpop.gremlin.process.traversal.TraversalInterruptionComputerTest.shouldRespectThreadInterruptionInVertexStep(TraversalInterruptionComputerTest.java:100)

[ERROR] shouldRespectThreadInterruptionInVertexStep[expectInterruption{g_E_inV}]  Time elapsed: 18.822 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: is <true>
     but: was <false>
	at org.apache.tinkerpop.gremlin.process.traversal.TraversalInterruptionComputerTest.shouldRespectThreadInterruptionInVertexStep(TraversalInterruptionComputerTest.java:100)

SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/Users/smcquill/.m2/repository/org/slf4j/slf4j-log4j12/1.7.30/slf4j-log4j12-1.7.30.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/Users/smcquill/.m2/repository/ch/qos/logback/logback-classic/1.1.3/logback-classic-1.1.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
[INFO] Running org.janusgraph.blueprints.cql.process.CQLMultiQueryProcessTest
[WARNING] Tests run: 785, Failures: 0, Errors: 0, Skipped: 19, Time elapsed: 11,985.447 s - in org.janusgraph.blueprints.cql.process.CQLMultiQueryProcessTest
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/Users/smcquill/.m2/repository/org/slf4j/slf4j-log4j12/1.7.30/slf4j-log4j12-1.7.30.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/Users/smcquill/.m2/repository/ch/qos/logback/logback-classic/1.1.3/logback-classic-1.1.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
[INFO] Running org.janusgraph.blueprints.cql.process.CQLProcessTest
[WARNING] Tests run: 785, Failures: 0, Errors: 0, Skipped: 19, Time elapsed: 12,251.599 s - in org.janusgraph.blueprints.cql.process.CQLProcessTest
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/Users/smcquill/.m2/repository/org/slf4j/slf4j-log4j12/1.7.30/slf4j-log4j12-1.7.30.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/Users/smcquill/.m2/repository/ch/qos/logback/logback-classic/1.1.3/logback-classic-1.1.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
[INFO] Running org.janusgraph.blueprints.cql.structure.CQLStructureTest
[WARNING] Tests run: 934, Failures: 0, Errors: 0, Skipped: 263, Time elapsed: 14,348.837 s - in org.janusgraph.blueprints.cql.structure.CQLStructureTest
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TraversalInterruptionComputerTest.shouldRespectThreadInterruptionInVertexStep:100 
Expected: is <true>
     but: was <false>
[ERROR]   TraversalInterruptionComputerTest.shouldRespectThreadInterruptionInVertexStep:100 
Expected: is <true>
     but: was <false>
[ERROR]   TraversalInterruptionComputerTest.shouldRespectThreadInterruptionInVertexStep:100 
Expected: is <true>
     but: was <false>
[ERROR]   TraversalInterruptionComputerTest.shouldRespectThreadInterruptionInVertexStep:100 
Expected: is <true>
     but: was <false>
[INFO] 
[ERROR] Tests run: 3111, Failures: 4, Errors: 0, Skipped: 277
```
But there does appear to be a problem with the `shouldRespectThreadInterruptionInVertexStep` test.

porunov commented:
@smcquillan The docker container is running but I still get those errors. I didn't go deep into it yet. In my situation the build result of `mvn clean install -Dtest.skip.tp=false -DskipTests=true` looks like:
```
...
[ERROR]   StarGraphTest.shouldSerializeCorrectlyUsingGryo Multiple Failures (2 failures)
	org.testcontainers.containers.ContainerLaunchException: Container startup failed
	java.lang.NullPointerException: <no message>
[ERROR]   StarGraphTest.shouldValidateThatOriginalAndStarVerticesHaveTheSameTopology Multiple Failures (2 failures)
	org.testcontainers.containers.ContainerLaunchException: Container startup failed
	java.lang.NullPointerException: <no message>
[INFO] 
[ERROR] Tests run: 1798, Failures: 1794, Errors: 0, Skipped: 4
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] JanusGraph: Distributed Graph Database 0.6.0-SNAPSHOT SUCCESS [  1.235 s]
[INFO] JanusGraph-Driver: Gremlin-Driver Library for JanusGraph SUCCESS [  4.302 s]
[INFO] JanusGraph-Core: Core Library for JanusGraph ....... SUCCESS [  3.272 s]
[INFO] JanusGraph-Backend-testutils: Collection of Backend tests SUCCESS [  3.903 s]
[INFO] JanusGraph-InMemory: InMemory Backend for JanusGraph SUCCESS [24:46 min]
[INFO] janusgraph-server .................................. SUCCESS [  0.331 s]
[INFO] JanusGraph-Test: Test Suite for JanusGraph ......... SUCCESS [  1.349 s]
[INFO] JanusGraph-BerkeleyJE: Distributed Graph Database .. SUCCESS [54:19 min]
[INFO] JanusGraph-Hadoop: Module .......................... SUCCESS [  3.628 s]
[INFO] JanusGraph-CQL: Distributed Graph Database ......... FAILURE [ 21.425 s]
[INFO] JanusGraph-HBase: Parent Module .................... SKIPPED
[INFO] JanusGraph-HBase: Version-independent Core ......... SKIPPED
[INFO] JanusGraph-HBase: HBase Server Used by Maven Testing SKIPPED
[INFO] JanusGraph-HBase: 1.x and 2.x Compatibility Shim ... SKIPPED
[INFO] JanusGraph-HBase: Universal binary ................. SKIPPED
[INFO] JanusGraph-Bigtable: Google Cloud Bigtable driver .. SKIPPED
[INFO] JanusGraph-ElasticSearch: Distributed Indexing Support SKIPPED
[INFO] JanusGraph-Lucene: Indexing Support ................ SKIPPED
[INFO] JanusGraph-Solr: Distributed Indexing Support ...... SKIPPED
[INFO] JanusGraph-All: Complete JanusGraph Distribution ... SKIPPED
[INFO] JanusGraph-Dist: Tar and Zip Archives .............. SKIPPED
[INFO] JanusGraph-Doc: Manual for JanusGraph .............. SKIPPED
[INFO] JanusGraph-Examples: Examples for JanusGraph ....... SKIPPED
[INFO] Example-Common: Common Graph Code for Examples ..... SKIPPED
[INFO] Example-BerkeleyJE: BerkeleyJE Storage, Lucene Index SKIPPED
[INFO] Example-Cql: Cassandra CQL Storage, Elasticsearch Index SKIPPED
[INFO] Example-HBase: HBase Storage, Solr Index ........... SKIPPED
[INFO] Example-RemoteGraph: Example with RemoteGraph ...... SKIPPED
[INFO] Example-TinkerGraph: Example with TinkerGraph 0.6.0-SNAPSHOT SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:19 h
[INFO] Finished at: 2020-08-24T00:31:47+03:00
[INFO] ------------------------------------------------------------------------
```

I guess it might be because I mix datastax driver 4 and 3 in the code but I'm not sure yet. Maybe I didn't correctly wrote updates for the cql-hadoop part which is using datastax driver 3.x. Not sure yet why there are problems to start a container even so the docker is running. Will try to figure our it later

smcquillan commented:
@porunov I found some time to look at this again and this time tried to run the tinker pop tests on Linux (my previous attempt had been on my MacBook). So I ran `mvn clean install -Dtest.skip.tp=false -DskipTests=true` and it took 12 hours but looks like all the tests passed:
```
# grep run build.log 
[INFO] No tests to run.
[INFO] No tests to run.
[WARNING] Tests run: 773, Failures: 0, Errors: 0, Skipped: 142, Time elapsed: 1,001.366 s - in org.janusgraph.blueprints.inmemory.process.InMemoryJanusGraphComputerTest
[WARNING] Tests run: 785, Failures: 0, Errors: 0, Skipped: 19, Time elapsed: 186.012 s - in org.janusgraph.blueprints.inmemory.process.InMemoryJanusGraphProcessTest
[WARNING] Tests run: 785, Failures: 0, Errors: 0, Skipped: 19, Time elapsed: 185.676 s - in org.janusgraph.blueprints.inmemory.process.InMemoryMultiQueryJanusGraphProcessTest
[WARNING] Tests run: 934, Failures: 0, Errors: 0, Skipped: 266, Time elapsed: 134.604 s - in org.janusgraph.blueprints.inmemory.structure.InMemoryJanusGraphStructureTest
[WARNING] Tests run: 3111, Failures: 0, Errors: 0, Skipped: 280
[WARNING] Tests run: 763, Failures: 0, Errors: 0, Skipped: 142, Time elapsed: 1,321.411 s - in org.janusgraph.blueprints.process.BerkeleyJanusGraphComputerTest
[WARNING] Tests run: 775, Failures: 0, Errors: 0, Skipped: 19, Time elapsed: 534.518 s - in org.janusgraph.blueprints.process.BerkeleyJanusGraphProcessTest
[WARNING] Tests run: 775, Failures: 0, Errors: 0, Skipped: 19, Time elapsed: 537.533 s - in org.janusgraph.blueprints.process.BerkeleyMultiQueryJanusGraphProcessTest
[WARNING] Tests run: 934, Failures: 0, Errors: 0, Skipped: 263, Time elapsed: 665.487 s - in org.janusgraph.blueprints.structure.BerkeleyJanusGraphStructureTest
[WARNING] Tests run: 3081, Failures: 0, Errors: 0, Skipped: 277
[WARNING] Tests run: 773, Failures: 0, Errors: 0, Skipped: 142, Time elapsed: 4,083.679 s - in org.janusgraph.blueprints.cql.process.CQLComputerTest
[WARNING] Tests run: 785, Failures: 0, Errors: 0, Skipped: 19, Time elapsed: 4,448.402 s - in org.janusgraph.blueprints.cql.process.CQLMultiQueryProcessTest
[WARNING] Tests run: 785, Failures: 0, Errors: 0, Skipped: 19, Time elapsed: 4,485.836 s - in org.janusgraph.blueprints.cql.process.CQLProcessTest
[WARNING] Tests run: 934, Failures: 0, Errors: 0, Skipped: 263, Time elapsed: 4,142.508 s - in org.janusgraph.blueprints.cql.structure.CQLStructureTest
[WARNING] Tests run: 3111, Failures: 0, Errors: 0, Skipped: 277
[WARNING] Tests run: 773, Failures: 0, Errors: 0, Skipped: 142, Time elapsed: 4,905.588 s - in org.janusgraph.blueprints.process.HBaseComputerTest
[WARNING] Tests run: 785, Failures: 0, Errors: 0, Skipped: 19, Time elapsed: 5,902.264 s - in org.janusgraph.blueprints.process.HBaseMultiQueryProcessTest
[WARNING] Tests run: 785, Failures: 0, Errors: 0, Skipped: 19, Time elapsed: 6,014.822 s - in org.janusgraph.blueprints.process.HBaseProcessTest
[WARNING] Tests run: 934, Failures: 0, Errors: 0, Skipped: 263, Time elapsed: 3,768.307 s - in org.janusgraph.blueprints.structure.HBaseStructureTest
[WARNING] Tests run: 3111, Failures: 0, Errors: 0, Skipped: 277
```
And the summary:
```
[INFO] Reactor Summary:
[INFO]
[INFO] JanusGraph: Distributed Graph Database ............ SUCCESS [1.308s]
[INFO] JanusGraph-Driver: Gremlin-Driver Library for JanusGraph  SUCCESS [5.809s]
[INFO] JanusGraph-Core: Core Library for JanusGraph ...... SUCCESS [5.769s]
[INFO] JanusGraph-Backend-testutils: Collection of Backend tests  SUCCESS [5.450s]
[INFO] JanusGraph-InMemory: InMemory Backend for JanusGraph  SUCCESS [25:17.313s]
[INFO] janusgraph-server ................................. SUCCESS [0.963s]
[INFO] JanusGraph-Test: Test Suite for JanusGraph ........ SUCCESS [2.383s]
[INFO] JanusGraph-BerkeleyJE: Distributed Graph Database . SUCCESS [51:07.747s]
[INFO] JanusGraph-Hadoop: Module ......................... SUCCESS [4.911s]
[INFO] JanusGraph-CQL: Distributed Graph Database ........ SUCCESS [4:46:19.053s]
[INFO] JanusGraph-HBase: Parent Module ................... SUCCESS [0.541s]
[INFO] JanusGraph-HBase: Version-independent Core ........ SUCCESS [9.528s]
[INFO] JanusGraph-HBase: HBase Server Used by Maven Testing  SUCCESS [5.252s]
[INFO] JanusGraph-HBase: 1.x and 2.x Compatibility Shim .. SUCCESS [5:43:34.673s]
[INFO] JanusGraph-HBase: Universal binary ................ SUCCESS [0.303s]
[INFO] JanusGraph-Bigtable: Google Cloud Bigtable driver . SUCCESS [0.191s]
[INFO] JanusGraph-ElasticSearch: Distributed Indexing Support  SUCCESS [2.030s]
[INFO] JanusGraph-Lucene: Indexing Support ............... SUCCESS [0.899s]
[INFO] JanusGraph-Solr: Distributed Indexing Support ..... SUCCESS [3.087s]
[INFO] JanusGraph-All: Complete JanusGraph Distribution .. SUCCESS [0.423s]
[INFO] JanusGraph-Dist: Tar and Zip Archives ............. SUCCESS [28.974s]
[INFO] JanusGraph-Doc: Manual for JanusGraph ............. SUCCESS [0.857s]
[INFO] JanusGraph-Examples: Examples for JanusGraph ...... SUCCESS [0.265s]
[INFO] Example-Common: Common Graph Code for Examples .... SUCCESS [0.284s]
[INFO] Example-BerkeleyJE: BerkeleyJE Storage, Lucene Index  SUCCESS [0.313s]
[INFO] Example-Cql: Cassandra CQL Storage, Elasticsearch Index  SUCCESS [0.290s]
[INFO] Example-HBase: HBase Storage, Solr Index .......... SUCCESS [0.273s]
[INFO] Example-RemoteGraph: Example with RemoteGraph ..... SUCCESS [0.277s]
[INFO] Example-TinkerGraph: Example with TinkerGraph ..... SUCCESS [0.220s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 11:47:39.819s
[INFO] Finished at: Tue Sep 08 21:41:43 PDT 2020
[INFO] Final Memory: 173M/1495M
[INFO] ------------------------------------------------------------------------
```
I've only seen the `org.testcontainers.containers.ContainerLaunchException: Container startup failed` error you've been seeing when I ran without docker running. I know you are running docker, but suspect there's something related to docker not right in your environment.

porunov commented:
@smcquillan Thank you for the investigation! Interesting, maybe it's really related to my server environment (even so I double checked that Docker is running and works as expected). Today night I will start TinkerPop tests in a different environment and will see how it goes.

porunov commented:
@smcquillan you were right. The problem was in my environment. TP CQL tests passed in another environment.
I've added `storage.cql.metrics` which now configures metrics the user wants to receive. I didn't check if that works yet. Will check later but overall I think this PR close to be finished (if not already finished).

porunov commented:
@marco-scoppetta I found several things which you might be interested:
1) CompactionStrategy `SizeTieredCompactionStrategy` will use `leveledCompactionStrategy`. This is a bug which still persists in grakn. https://github.com/graknlabs/grakn/blob/cf4de7f936bec2a06e63e4dc3d74543a40e86e99/graph/diskstorage/cql/CQLKeyColumnValueStore.java#L273 
2) I know that the DataStax Cassandra Driver 4.x now uses `class` to specify `compression` algorithm. Thus, you are correct by using it here: https://github.com/graknlabs/grakn/blob/cf4de7f936bec2a06e63e4dc3d74543a40e86e99/graph/diskstorage/cql/CQLKeyColumnValueStore.java#L262 but due to the bug on ScyllaDB side https://github.com/scylladb/scylla/issues/4200 it still expects `sstable_compression` (which is used in DataStax Cassandra Driver 3.x) instead of `class`. Thus, I changed `class` to `sstable_compression` in this PR so that it would be compatible with ScyllaDB as well.

porunov commented:
The PR is now ready for review. Description and information about conducted tests are included into the [PR description](https://github.com/JanusGraph/janusgraph/pull/2169#issue-448959364). Appreciate any reviews.

li-boxuan commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:
```diff
@@ -434,43 +444,43 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
+            if(currentResultSet.isFullyFetched()){
+                return false;
+            }
+            updateResultSet();
```
Do you know when this line will be executed? Codecov shows this is not covered. By testing locally, I created a graph with half a million vertices and did a full scan, but still, this line was not executed.

porunov replied:
Hmm. Interesting. This line should be executed when `currentPageIterator` has no any elements on this page but there are still some pages available i.e. `!currentPageIterator.hasNext() && !currentResultSet.isFullyFetched()`

porunov replied:
Did you have `1,000,000` vertices in `currentPageIterator`?

porunov replied:
What was your `storeManager.getPageSize()`?

porunov replied:
Interesting. This line isn't reachable for me as well.

porunov replied:
I will check that again. Looks like pages are fetched automatically but as far as I remember it wasn't true for the new driver. Maybe something changed or I missed something and pages are fetched automatically in the iterator. I will check that. Nevertheless this logic shouldn't break anything but if this logic is redundant we should remove it to simplify it.

porunov replied:
Okay, I checked, indeed DSE driver automatically fetches pages but now I see why this iterator was implemented. It has a comment above that automatic DSE pagination could lead to OOM because it was storing the whole history. Thus, we were adding another level of pagination. I will check if it is the issue still and if so will fix this part of the logic. Thank you for catching this!

porunov replied:
`ExecutionInfo` is stored for each new page as seen here:
https://github.com/datastax/java-driver/blob/c225588811a85e25d98790ce4b4f399e0ca066e4/core/src/main/java/com/datastax/dse/driver/internal/core/graph/MultiPageGraphResultSet.java#L108

So, potentially it may result in OOM. Thus, I'm changing the logic of this iterator back to forcing new request per page as it was before. It's great that you noticed this issue @li-boxuan ! Thanks!

porunov replied:
Fixed with explanations here: https://github.com/JanusGraph/janusgraph/pull/2169#issuecomment-730041434

li-boxuan commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLConfigOptions.java:
```diff
@@ -283,6 +284,175 @@
+        NETTY,
+        "timer-tick-duration",
+        "The timer tick duration in milliseconds. This is how frequent the timer should wake up to check for timed-out tasks " +
+            "or speculative executions. See DataStax Java Driver option " +
```
nitpick: missing backquote around `DefaultDriverOption.NETTY_TIMER_TICK_DURATION.getPath()`. Same for next config option.

porunov commented:
@li-boxuan I've force pushed changes here: https://github.com/JanusGraph/janusgraph/compare/672bfb0c37039deefedeb7522afc761803b00e81..eba708f181a22ce9e7698f7a1d8d449c0d132e45

I fixed `CQLPagingIterator` to relay on amount of fetched elements (as it was) rather then on internal automatic page retrieval (to not reintroduce a possible OOM issue).
I've fixed documentation comments.
Also, I changed `bind` to `boundStatementBuilder` because `bind` creates a statement which is recreated on every mutation (redundant objects). `boundStatementBuilder` creates a statement only during `build` call.
Also, I don't see the point of a `Supplier` anymore which was passed to `CQLPagingIterator` thus I unwrapped the statement creation function and changed that to pass a statement (initial statement) to `CQLPagingIterator`.
As I said before, right now every `BoundStatement` mutation results in a new `BoundStatement` thus `boundStatement.setPagingState(lastPagingState)` will create a new `BoundStatement`, that's why I don't see any value in `Supplier` which was used before.

li-boxuan commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:
```diff
@@ -434,43 +443,48 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
+                return currentPageIterator.hasNext();
+            }
+            if(currentResultSet.isFullyFetched()){
+                return false;
```
nice, but seems now this line is unreachable.

li-boxuan commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:
```diff
@@ -434,43 +443,48 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
+                return currentPageIterator.hasNext();
+            }
+            if(currentResultSet.isFullyFetched()){
+                return false;
```
Nice! Just a nitpick:
Now this line seems unreachable. See codecov: https://app.codecov.io/gh/JanusGraph/janusgraph/compare/2169/diff

Also sample test code:

```java
    @Test
    public void testPagingIterator() {
        clopen(option(PAGE_SIZE), 10);
        int numVertices = 10;
        for (int i = 0; i < numVertices; i++) {
            graph.addVertex().property("prop", i);
        }
        graph.tx().commit();

        List<Object> values = graph.traversal().V().values("prop").toList();
        assertEquals(numVertices, values.size());
        assertEquals(numVertices, new HashSet<>(values).size());
    }
```

Seems the underlying Datastax driver needs to execute the query once more to know that the current result set is fully fetched. Thus this if-statement seems to be useless.

li-boxuan replied:
Interestingly, this line is reachable and is necessary when paginatedResultSize is negative. Seems Datastax driver allows negative value, simply ignores it, and uses the default value. If we don't have this if-statement, our program will go into an endless loop.

li-boxuan replied:
The following test case fails under CQLGraphTest but can pass InMemoryGraphTest & BerkeleyGraphTest:
```java    
    @Test
    public void testPagingIterator() {
        clopen(option(PAGE_SIZE), 10);
        int numVertices = 10;
        for (int i = 0; i < numVertices; i++) {
            graph.addVertex().property("prop", i);
        }
        graph.tx().commit();

        List<Object> values = graph.traversal().V().values("prop").toList();
        assertEquals(numVertices, values.size());
        assertEquals(numVertices, new HashSet<>(values).size());
    }
```
didn't check thoroughly but I guess it's because the implementation here assumes paging size is non-negative.

porunov replied:
Oh, didn't think about this edge case. For some reason I thought that negative value isn't allowed. I will fix that

porunov commented:
@li-boxuan changed the implementation not to relay on PageSize at all. https://github.com/JanusGraph/janusgraph/compare/eba708f181a22ce9e7698f7a1d8d449c0d132e45..70321725c0cd1be7a40ceedd36a43a94127d0eb1

What do you think about it?

porunov commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:
```diff
@@ -414,63 +423,50 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
+            if(dseStoredExecutionInfos.size()>1){
+                if(dseStoredExecutionInfos.size()==2){
+                    dseStoredExecutionInfos.set(0, dseStoredExecutionInfos.remove(1));
+                } else {
+                    ExecutionInfo lastExecutionInfo = dseStoredExecutionInfos.get(dseStoredExecutionInfos.size()-1);
+                    dseStoredExecutionInfos.clear();
+                    dseStoredExecutionInfos.add(lastExecutionInfo);
+                }
```
This part should be unreachable in most cases. The possible case I think about when this part is actually reachable is when we have `storage.page-size=1` and call `next() & hasNext() & next()`. In such case it's possible that during the last `next()` operation our `dseStoredExecutionInfos.size()` will be equal to `3`. In normal scenarios when we always call `hasNext()` before `next()` we shouldn't reach this `else` case.

porunov commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:
```diff
@@ -414,63 +423,50 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
-            final Statement boundStmnt = statementSupplier.get();
-            if (lastPagingState != null) {
-                boundStmnt.setPagingStateUnsafe(lastPagingState);
+            if(dseStoredExecutionInfos.size()>1){
```
Current DSE `ResultSet` implementations return mutable `ArrayList` when `getExecutionInfos()` is called. The only case when this list is Immutable is when `SinglePageResultSet` is used but we won't mutate that list because it always has the size of `1`. Thus, it's not possible to enter this `if` statement with that Immutable list.

li-boxuan commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:
```diff
@@ -414,63 +423,50 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
-                boundStmnt.setPagingStateUnsafe(lastPagingState);
+            if(dseStoredExecutionInfos.size()>1){
+                if(dseStoredExecutionInfos.size()==2){
+                    dseStoredExecutionInfos.set(0, dseStoredExecutionInfos.remove(1));
```
What about `dseStoredExecutionInfos.remove(0)`? Actually, if we simply do `dseStoredExecutionInfos.remove(0)`, or use your code in the `else` branch, then we don't need two branches at all. Is there any specific reason to keep the `if(dseStoredExecutionInfos.size()==2)` condition here?

porunov replied:
`dseStoredExecutionInfos.remove(0)` - will remove only the first element. So, it there are 3 elements 2 elements will be left. So, in case we execute `next(); hasNext(); next(); hasNext(); next(); hasNext();` with page size `1` it will produce `5` elements in `dseStoredExecutionInfos` because on `next(); hasNext();` we are adding 2 elments but on `nest()` we are removing only 1 element.
`dseStoredExecutionInfos.remove(0)` and `dseStoredExecutionInfos.set(0, dseStoredExecutionInfos.remove(1))` are kind of the same if the `size == 2` but when we execute `dseStoredExecutionInfos.remove(1)` we are not triggering `System.arraycopy` because we removed the last element and when we execute `dseStoredExecutionInfos.remove(0)` we trigger that. From another hand `System.arraycopy` will copy just 1 element, thus it will be fast. I just thought that `dseStoredExecutionInfos.set(0, dseStoredExecutionInfos.remove(1))` would be several operations faster then `dseStoredExecutionInfos.remove(0)` but it shouldn't be much difference and I didn't really compare the performance. Just assumed

porunov replied:
Ah, I got it. I misunderstood second part. You are right. We don't need 2 branches if we use the code in `else` branch. I was just optimizing usage of unnecessary operations here but I think it just made the code more complex

porunov replied:
Updated the code as you suggested here: https://github.com/JanusGraph/janusgraph/compare/70321725c0cd1be7a40ceedd36a43a94127d0eb1..65cad01828f1a455cc51435f8efaf309111c0768

li-boxuan reviewed (commented):
The `CQLPagingIterator` part looks nice now 👍 
I checked the usage of `List<ExecutionInfo> infos` in the DataStax driver codebase and was surprised to find only the last ExecutionInfo was used internally. Probably there are other use cases, but since we don't have the need, your implementation should be good.

porunov commented:
@li-boxuan @farodin91 @smcquillan @marco-scoppetta @FlorianHockmann @rngcntr @twilmes @pluradj 
I have an intention to merge this PR by following a lazy consensus. I case anyone has an intention to review this PR and needs some time, please say, so that I would hold off on merging it.

li-boxuan reviewed (approved):
I went through all changes and look good to me 👍 
I didn't test them and cannot say I understand every piece though

smcquillan reviewed (approved):
I created a custom build of our application to make it use an updated `janusgraph-cql` with the `4.9.0` driver and ran it though our integration test suite. All tests passed so changes look  good to me.
